lxd/main_daemon.go

```diff
@@ -76,23 +76,42 @@ func (c *cmdDaemon) Run(cmd *cobra.Command, args []string) error {
 	chIgnore := make(chan os.Signal)
 	signal.Notify(chIgnore, unix.SIGHUP)
 
+	chForceShutdown := make(chan bool)
+	chDone := make(chan bool)
+
 	s := d.State()
 	select {
 	case sig := <-ch:
 		if sig == unix.SIGPWR {
-			logger.Infof("Received '%s signal', shutting down containers", sig)
+			logger.Infof("Received '%s signal', shutting down instances", sig)
 			d.Kill()
-			containersShutdown(s)
+			instancesShutdown(s)
 			networkShutdown(s)
 		} else {
 			logger.Infof("Received '%s signal', exiting", sig)
 			d.Kill()
 		}
 
 	case <-d.shutdownChan:
-		logger.Infof("Asked to shutdown by API, shutting down containers")
+		logger.Infof("Asked to shutdown by API, waiting for all operations to finish")
+		// Cancelling the context will make everyone aware that we're shutting down.
+		d.cancel()
+
+		go func() {
```

Comment:
@monstermunchkin actually looking at this more closely, why run this in a separate go routine at all? Instead just call `waitForOperations` to return (and ensure it does return when chForceShutdown is closed). That way you wouldn't need the `chDone` channel at all, as it is you're effectively making waitForOperations blocking anyway.